tutorial/lesson_16_rgb_generator.cpp

```diff
@@ -0,0 +1,190 @@
+// Halide tutorial lesson 16: RGB images and memory layouts
+
+// This lesson demonstrates how to feed Halide RGB images in
+// interleaved or planar format, and how to write code optimized for
+// each case.
+
+/* On linux or os x, you can compile and run it like so:
+
+g++ lesson_16_rgb_generator.cpp ../tools/GenGen.cpp -g -std=c++11 -fno-rtti -I ../include -L ../bin -lHalide -lpthread -ldl -o lesson_16_generate
+export LD_LIBRARY_PATH=../bin
+export DYLD_LIBRARY_PATH=../bin
+./lesson_16_generate -o . -f brighten_planar      target=host layout=planar
+./lesson_16_generate -o . -f brighten_interleaved target=host layout=interleaved
+./lesson_16_generate -o . -f brighten_either      target=host layout=either
+./lesson_16_generate -o . -f brighten_specialized target=host layout=specialized
+g++ lesson_16_rgb_run.cpp brighten_*.o -ldl -lpthread -o lesson_16_run
+./lesson_16_run
+
+*/
+
+// If you have the entire Halide source tree, you can also build it by
+// running:
+//    make tutorial_lesson_16_rgb_run
+// in a shell with the current directory at the top of the halide
+// source tree.
+
+#include "Halide.h"
+#include <stdio.h>
+
+using namespace Halide;
+
+// We will define a generator that brightens an RGB image.
+class Brighten : public Halide::Generator<Brighten> {
+public:
+    // We declare a three-dimensional input image. The first two
+    // dimensions will be x, and y, and the third dimension will be
+    // the color channel.
+    ImageParam input{UInt(8), 3, "input"};
+
+    // We will compile this generator in several ways to accept
+    // several different memory layouts for the input and output. This
+    // is a good use of a GeneratorParam (see lesson 15).
+    enum class Layout { Planar, Interleaved, Either, Specialized };
+    GeneratorParam<Layout> layout{"layout",
+            // default value
+            Layout::Planar,
+            // map from names to values
+                {{ "planar",        Layout::Planar },
+                 { "interleaved",   Layout::Interleaved },
+                 { "either",        Layout::Either },
+                 { "specialized",   Layout::Specialized }}};
+
+    // We also declare a scalar parameter to control the amount of
+    // brightening.
+    Param<uint8_t> offset{"offset"};
+
+    // Declare our Vars
+    Var x, y, c;
+
+    Func build() {
+        // Define the Func.
+        Func brighter("brighter");
+        brighter(x, y, c) = input(x, y, c) + offset;
+
+        // Schedule it.
+        brighter.vectorize(x, 16);
+
+        // We will compile this pipeline to handle memory layouts in
+        // several different ways, depending on the 'layout' generator
+        // param.
+        if (layout == Layout::Planar) {
+            // This pipeline as written will work with planar images only,
```

Comment:
I'd like this to indicate that the same schedule/code works for line by line layout -- a row of red, a row of green, a row of blue. (This layout may offer better locality than completely planar.)

It may also be worth discussion constraints such as stride == extent, stride == aligned-extent, stride and extent == a constant value. Oh, and of course anything involving alignment has to cover the mins as well so this might be a good place to mention that as an aside too. These may be a distraction from what this tutorial is about, but I'm thinking it is worth noting their existence so folks know this kind of thing can be done.

Consider adding something like these visual diagram comments as well:

// RRRRRRRR
// RRRRRRRR
// RRRRRRRR
// GGGGGGGG
// GGGGGGGG
// GGGGGGGG
// BBBBBBBB
// BBBBBBBB
// BBBBBBBB

// RGBRGBRGBRGBRGBRGBRGBRGBRGBRGBRGBRGB
// RGBRGBRGBRGBRGBRGBRGBRGBRGBRGBRGBRGB
// RGBRGBRGBRGBRGBRGBRGBRGBRGBRGBRGBRGB

// RRRRRRRR
// GGGGGGGG
// BBBBBBBBB
// RRRRRRRR
// GGGGGGGG
// BBBBBBBBB
// RRRRRRRR
// GGGGGGGG
// BBBBBBBBB

Or perhaps we should draw a nice visual using something with more resolution than ASCII art and link to it from here.